db/migrate/20160307175910_migrate_hs_tags.rb

```diff
@@ -0,0 +1,142 @@
+    display_exercises.each do |exercise|
+      ExerciseTag.create!(exercise: exercise, tag: tl_multiple_choice_tag)
+    end
+
+    rec_sensitive_tag = Tag.find_by(name: 'display-rec-sensitive') # Unused
+    new_tag rec_sensitive_tag, 'display:rec-sensitive'
+
+    # Tagging legend changes
+    tl_id_tags = Tag.where{name.like 'id:%'} # Unused (CC does not use exercise ID's)
+    tl_id_tags.each{ tag.update_attribute :name, tag.name.gsub('id:', 'exid:') }
+
+    tl_type_tags = Tag.where{name.like 'ost-type:%'} # Unused (CC uses all exercises)
+    tl_type_tags.each{ |tag| tag.update_attribute :name, tag.name.gsub('ost-type:', 'type:') }
+
+    # Type tags
+    inbook_tag = Tag.find_by(name: 'inbook-yes') # Unused
+    new_tag inbook_tag, 'filter-type:inbook'
+    new_tag inbook_tag, 'type:conceptual-or-recall'
+
+    old_practice_tag = Tag.find_by(name: 'os-practice-problems') # Used by Tutor
+    new_tag old_practice_tag, 'type:practice'
+
+    old_concepts_tag = Tag.find_by(name: 'os-practice-concepts') # Used by Tutor
+    new_tag old_concepts_tag, 'type:conceptual'
+
+    conceptual_tag = Tag.find_or_create_by(name: 'type:conceptual')
+    practice_tag = Tag.find_or_create_by(name: 'type:practice')
+
+    old_cr_tag = Tag.find_by(name: 'ost-chapter-review') # Used by Tutor
+    chapter_review_exercise_tags = old_cr_tag.try(:exercise_tags)
+                                             .try(:preload, exercise: :tags) || []
+    chapter_review_exercise_tags.each do |et|
+      tag = et.exercise.tags.map(&:name).include?('concept') ? conceptual_tag : practice_tag
+      ExerciseTag.create!(exercise: et.exercise, tag: tag)
+    end
+    new_tag old_cr_tag, 'filter-type:chapter-review'
+
+    old_tp_tag = Tag.find_by(name: 'ost-test-prep') # Unused
+    new_tag old_tp_tag, 'type:practice'
+    new_tag old_tp_tag, 'filter-type:test-prep'
+  end
+
+  def down
+    raise ActiveRecord::IrreversibleMigration, 'Just restore the backup...'
+  end
+
+  protected
+
+  # Creates a new tag and associates it with the same records as the given tag
+  def new_tag(tag, name)
+    return if tag.nil?
+
+    @tags ||= {}
+    @tags[name] ||= Tag.find_or_create_by(name: name)
+    tag.exercise_tags.each{ |et| ExerciseTag.create!(exercise: et.exercise, tag: @tags[name]) }
+  end
+
+  # Puts the module UUID's from a CNX archive JSON hash into the cnx_id_map
+  def map_collection(hash, cnx_id_map, chapter_number = 0)
+    contents = hash['contents']
+    chapter_number += 1 if contents.none?{ |hash| hash['id'] == 'subcol' }
+
+    page_number = nil
+    contents.each do |entry|
+      if entry['id'] == 'subcol'
+        chapter_number = map_collection(entry, cnx_id_map, chapter_number)
+      else
+        page_number ||= entry['title'].start_with?('Introduction') ? 0 : 1
+        cnx_id_map[chapter_number][page_number] = entry['id'].split('@').first
+        page_number += 1
+      end
+    end
+
+    return chapter_number
+  end
```

Comment:
When I talked about writing a spec, I'm actually talking about all of the code not just this bit at the end :-)